### PR TITLE
Use Postgres instead of JSON

### PR DESCRIPTION
**Pull request by mplewis**
Hey @tylrtrmbl! This is halfway there.

Things we need to do:
- Move redirects to Postgres
- Update the redirects controller to grab data from Postgres
- Move server connection stuff to `config.py`
- ???
- Profit!

### CODE REVIEW

**Comment by mplewis** [comment]
Psst. This closes #143.

File: projects_controller.py [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
+PROJECTS_KEY_ORDER = ['id', 'name', 'slug', 'description', 'photo_url', 'past_project']
+GET_PROJECTS_QUERY = 'SELECT project_id, name, slug, description, photo_url, past_project FROM projects ORDER BY display_order;'
+
+GET_PROJECT_PHOTOS_QUERY = 'SELECT photo_url FROM project_photos WHERE project_id = %s ORDER BY display_order;'
+
+GET_PROJECT_NEEDS_QUERY = 'SELECT need_text FROM project_needs WHERE project_id = %s ORDER BY display_order;'
+
+PROJECT_LEADER_KEY_ORDER = ['name', 'phone', 'email', 'bio', 'photo_url']
+GET_PROJECT_LEADERS_QUERY = (
+    'SELECT name, phone, email, bio, photo_url FROM leaders WHERE leader_id IN '
```
Comment (taylortrimble) [inline projects_controller.py]: I feel like there is some way to do this better with a `JOIN`

File: projects_controller.py [hunk 2]
```diff
@@ -28,37 +39,64 @@ def get_all_projects(self):
...
-        all_files = os.listdir(self.projects_dir)
-        json_files = [f for f in all_files if f.endswith('.json')]
-        for file_name in json_files:
-            full_file_name = os.path.join(self.projects_dir, file_name)
-            with open(full_file_name, 'r') as f:
-                data = json.load(f)
-        
-            project_key = os.path.splitext(file_name)[0]
+
+        from pprint import pprint
```
Comment (taylortrimble) [inline projects_controller.py]: I've only started seeing this recently, so I wonder: is it Pythonic to keep these includes up top, or just better in some other way like being upfront with imports?

File: projects_controller.py [hunk 3]
```diff
@@ -28,37 +39,64 @@ def get_all_projects(self):
...
+
+        try:
+            con = psycopg2.connect(host='localhost', database='twnet', user='mplewis')
```
Comment (taylortrimble) [inline projects_controller.py]: Of course moving out to config. :)

File: templates/index.html [hunk 4]
```diff
@@ -116,9 +116,7 @@ <h4 class="text-muted"> Join us Fridays at 5pm in Keller 3-230</h4>
 {% block html_post_body %}
 
 <script type="text/javascript">
-
```
Comment (taylortrimble) [inline templates/index.html:118]: Aww I liked these whitespaces, haha

File: templates/macros/form.html [hunk 5]
```diff
@@ -29,7 +29,7 @@
 {% macro textarea(id, label, placeholder, size, rows) %}
 
 <div class="form-group{% if errors[id] %} has-error{% endif %}">
-  <label for="{{ id }}">Project Description</label>
```
Comment (taylortrimble) [inline templates/macros/form.html:32]: Good catch

File: templates/project.html [hunk 6]
```diff
@@ -10,17 +10,17 @@
 {% block html_body %}  
   
 <div class="row text-center">
-    <h1>{{ project_data.project_title }}</h1>
+    <h1>{{ project_data.name }}</h1>
```
Comment (taylortrimble) [inline templates/project.html:12]: Liking the new keys (`name`, `slug`) a lot better.

File: projects_controller.py (hunk 2 above)
Comment (mplewis) [inline projects_controller.py]: The pythonic way to do it is to not use pprint in production :D 

Please kill this line and its dependencies, I didn't mean to leave them in.

File: templates/project.html (hunk 6 above)
Comment (mplewis) [inline templates/project.html:12]: Yeah, I don't think we need to be as verbose as we were.

File: projects_controller.py (hunk 1 above)
Comment (mplewis) [inline projects_controller.py]: I bet there is. I sent you the DB backup so you can toy with it!

Try Postgres.app, pgAdmin3, Induction, and any combination of the three.

**Comment by taylortrimble** [comment]
Big Big Question: now that we are moving to a DB, are we still going to cache on our side? I think if we're the only client of the DB then it's okay, but we would need to be upfront about **_THIS WEBSITE IS HOSTED FROM ONE APP ONLY OKAY??_**.

I bet we'd see performance boost (aka stay good), but we should be sure it doesn't make consistency ugly.

File: projects_controller.py (hunk 2 above)
Comment (taylortrimble) [inline projects_controller.py]: Lol, okay, I'll let you do this :)

File: templates/project.html (hunk 6 above)
Comment (taylortrimble) [inline templates/project.html:12]: st-st-stutterin'!

`project_data.project_title`

**Comment by mplewis** [comment]
I think caching is good. How do you feel about a cron job to recache every 60 seconds? Or 300, etc.

**Comment by taylortrimble** [comment]
Like a timer inside the app? Personally, I'd rather hit it every time or guarantee in-memory and in-db consistency

**Comment by mplewis** [comment]
How about: Since we're using in-app editors for DB data now, just hit an internal recache link whenever data is updated?

**Comment by taylortrimble** [comment]
Bingo

**Comment by taylortrimble** [comment]
I honestly still haven't been through this yet. Update to the control panel, etc, coming soon?

**Comment by taylortrimble** [comment]
Wait, this isn't the control panel PR is it? Just the rest of the data and config stuff (which might need to become config.json, I think we mentioned?) 

Cool. I'll maybe test soon if I need break from Fido.

**Comment by mplewis** [comment]
Correct on all counts!

**Comment by taylortrimble** [comment]
This code and the dump I have are out of sync to a point that's annoying to fix. Can I get a recent dump?

File: projects_controller.py [hunk 7]
```diff
@@ -1,10 +1,21 @@
-import json
-import os
+import psycopg2
+
+PROJECTS_KEY_ORDER = ['id', 'name', 'slug', 'description', 'photo_url', 'past_project']
+GET_PROJECTS_QUERY = 'SELECT project_id, name, slug, description, photo_url, past_project FROM projects ORDER BY display_order;'
```
Comment (taylortrimble) [inline projects_controller.py]: The out-of-date dump I have calls `project_id` just `id`, which I like a lot better.

File: projects_controller.py (hunk 3 above)
Comment (taylortrimble) [inline projects_controller.py]: Also we should probably initialize this once at the head of the app and keep it.

File: projects_controller.py (hunk 7 above)
Comment (taylortrimble) [inline projects_controller.py]: Can we generate this query from the key order? So they don't get out of sync like they are now?

```
def get_projects_query(keys):
  return 'SELECT %s FROM ...' % keys.toStringList

def unpack_db_results(keys, results):
  dict = {}
  for xrange thingy load dem keys into dicts by index matching
```

And then it's all in sync and works and shit and unpacking can happen in one line in the app.

**Comment by mplewis** [comment]
Added some commits. Please review and :sushi: or don't.

Tested and works on my machine.

Remind me of the current state of `config.py` please?

**Comment by mplewis** [comment]
Current (as of this second) Postgres dump is available at http://z.kesdev.com/code/1C1547083147

**Comment by taylortrimble** [comment]
I like everything but the stuttery `id`s! Not a fan of `leader_id` in the leader table, can just be `id`.

Do you also have the blank database generator, your plain text hand written db setup SQL, pushed somewhere?

**Comment by mplewis** [comment]
Here's the current create SQL script:
http://z.kesdev.com/code/2X070u2E381J
I don't have this pushed somewhere yet. Suggestions as to where I should push this?

Re: stutter IDs: This is actually a good thing for semantics IMO.
Check out `twnet-create.sql` above.

``` SQL
CREATE TABLE project_needs (
    project_need_id serial,
    project_id integer
        NOT NULL
        REFERENCES projects
        ON DELETE CASCADE,
    need_text text,
    display_order integer,
    PRIMARY KEY (project_need_id)
);
```

See the `REFERENCES projects` line? Since the `project_needs` column name `project_id` matches the `projects` column name `project_id`, we can shorten the command from the following:

``` SQL
        REFERENCES projects (project_id)
```

(or, if we were using just `id`)

``` SQL
        REFERENCES projects (id)
```

to:

``` SQL
        REFERENCES projects
```

I also like it because there's only one `project_id` in the database, not a couple `project_id`s and some `id`s that go with them. So the name is DB consistent.

Further reading: [http://www.postgresql.org/docs/8.2/static/ddl-constraints.html#DDL-CONSTRAINTS-FK](Postgres docs)

**Comment by taylortrimble** [comment]
> Here's the current create SQL script:
> http://z.kesdev.com/code/2X070u2E381J
> I don't have this pushed somewhere yet. Suggestions as to where I should push this?

Let's make it part of the repo!

> Re: stutter IDs: This is actually a good thing for semantics IMO.

Yeah, I'm on board with this now too.

I'll more thoroughly review now!

**Comment by mplewis** [comment]
> Let's make it part of the repo!

Done.

**Comment by taylortrimble** [comment]
I'm getting missing images for Ultralight, ODI. I'm guessing this is because we haven't merged in master and you got the data from the deployed site, right?

**Comment by taylortrimble** [comment]
We've still got some JSON lurking around: can we move the redirects to the db too?

**Comment by taylortrimble** [comment]
A few ideas I want to know if you're open to:

#### Object Key Name <==> Table Column Name

``` python
PROJECTS_KEY_ORDER = ['id', 'name', 'slug', 'description', 'photo_url', 'past_project']
GET_PROJECTS_QUERY = 'SELECT project_id, name, slug, description, photo_url, past_project FROM projects ORDER BY display_order;'
```

Becomes

``` python
PROJECTS_KEY_ORDER = ['project_id', 'name', 'slug', 'description', 'photo_url', 'past_project']
GET_PROJECTS_QUERY = 'SELECT %s FROM projects ORDER BY display_order;' % ', '.join(PROJECTS_KEY_ORDER)
```

I think it makes things a little DRYer, easy to add columns.

#### Generalizing object mapping from SQL rows

This happens twice:

``` python
for attr_num in xrange(len(project_raw)):
                    key_name = PROJECTS_KEY_ORDER[attr_num]
                    project_data[key_name] = project_raw[attr_num]
```

And this is a little hard to understand at first (but cool):

``` python
needs = [text_holder[0] for text_holder in cur.fetchall()]
```

We could have something like `object = make_object(obj_raw, key_order)` instead that generalizes to all those cases.

#### Same as above, and works for collections

`object = make_object_collection(objs_raw, key_order)`

We could _actually_ do this for projects, leaders, needs, AND photos. Do it for projects first, then add leaders and needs and photos to it after it's been `make_object_collection`ed first.

``` python
projects_data = make_object_collection(projects, key_order)
for project in projects_data:
    id = project['project_id']
    leaders_raw = # get the leaders by project id
    needs_raw = # get the leaders by project id
    photos_raw = # get the leaders by project id
    project['leaders'] = make_object_collection(photos_raw, PHOTOS_KEY_ORDER)
    project['needs'] = make_object_collection(photos_raw, PHOTOS_KEY_ORDER)
    project['photos'] = make_object_collection(photos_raw, PHOTOS_KEY_ORDER)
```

**Comment by taylortrimble** [comment]
#### Data storage
- Past projects will need a URL to redirect to upon completion (usually a blog post). We could perhaps get rid of `past_project` then and test for the existence of the redirect URL.
- Can we enforce that the `display_order` columns have unique values for each containing object for, say, `project_photo` relationships? I.e. Robomural doesn't have [1, 1, 3]?
- `create-tables.sql` is my new favorite file in this repo. Pretty.

#### Database connection

Let's open the connection to the database and keep it open forever. Opening connections like that each time ain't good for performance.

#### Writing

I don't see writing to the database in this guy.

#### A Bug and a Can of Worms (that look like snakes)

:bug:**:** If we don't have any past projects, we'll hit the db every time. The check `if not self.past_projects` will fail and we'll reload.

:snake:**:** What if we didn't do caching? We are making the assumption we are the only client on this, which is true, but it's kinda bad architecture. We're basically using the db as a backup of RAM.

It also solves our bug, and with an open connection I don't think the performance would be too bad! We could A/B of course.

**Comment by mplewis** [comment]
> Object Key Name <==> Table Column Name
> Generalizing object mapping from SQL rows
> Generalizing collection mapping from SQL rows

:+1: These are all great ideas. I'm loving DRY. What's the best way to document esoteric utility/generalizing functions like this?

> :bug: No Past Projects

Handled in commit `4258475`.

> :snake: Caching

Check out [this SO discussion](http://programmers.stackexchange.com/questions/154050/should-i-cache-the-data-or-hit-the-database).

Normally I'm in favor of caching but I don't think we'll actually run into enough speed issues to have to cache. If we cache wrong, we end up hurting our performance instead. How do you feel about putting off caching until we scale?

(Alternately, you could write a caching engine. I don't mind either way ;)

> DB Writing

There isn't any yet :(

Is that still in the scope of this PR? I'm moving existing functionality to Postgres.

> Redirect Slugs

I think this problem is complex enough to warrant its own PR. I'll iMessage before iCode.

**Comment by mplewis** [comment]
> Can we enforce that the `display_order` columns have unique values for each containing object for, say, `project_photo` relationships?

I don't think it matters all that much. `ORDER BY` is fine with duplicates. If we write our editor right, we should never run into a situation in which we submit `[1, 1, 3]`, etc.

I want to avoid constraints whenever not absolutely necessary. I don't like it as a dev when software tells me I can't do something because the last developer didn't think it made sense.

**Comment by mplewis** [comment]
Re: opening and closing a connection: We have to make the object properly destroyable.

http://stackoverflow.com/questions/865115/how-do-i-correctly-clean-up-a-python-object

I'm going to start using this pattern.

**Comment by mplewis** [comment]
Hey, I think making `GET_PROJECTS_QUERY = 'SELECT %s FROM projects ORDER BY display_order;'` may open us to injection. Can you think of a vector by which this might occur?

I'm not saying it's possible, I'm just saying using string interpolation and concatenation inside query strings is a **Bad Idea**.

**Comment by mplewis** [comment]
You know, despite those suggestions making the code DRYer, I think they complicate the code. Here's why:

We have a couple of uses of these cases. 2 of each:
- attr array --> attr dict
- single attr container array --> single attr array

In addition, the attr array --> attr dict is different in both cases:
- `for project_raw in projects:` sorts projects into a dict by slug keys
- `for leader_raw in leaders_raw:` puts leader dicts into an array, ordered by `display_order`

So, it's hard to DRY the attr array --> attr dict function because they're really two functions.

In addition, the single attr container array --> single attr array _is_ only two lines, and there are only two of them. And my personal guideline for DRY is:
- 3 or more = DRY it
- 2 or less = maybe

Wahtchu think?

Ima keep plugging at this. I think there are breakthroughs to be made.

**Comment by mplewis** [comment]
On XOR slug constraints: http://stackoverflow.com/questions/960556/sql-how-to-enforce-that-only-a-single-column-is-set-in-a-group-of-columns

**Comment by taylortrimble** [comment]
> Normally I'm in favor of caching but I don't think we'll actually run into enough speed issues to have to cache.

I agree, right now I think we are caching though and I want to nuke it! We "load". I think we should query the each time! Otherwise, we're unintentionally cacheing. Whoops.

**Comment by taylortrimble** [comment]
> Is that still in the scope of this PR?

Nope, nvm

**Comment by mplewis** [comment]
I'm down for killing caching. I like querying each time.

**Comment by taylortrimble** [comment]
Okay, #161 takes care of caching

**Comment by taylortrimble** [comment]
Our green merge button is back! Also, you have to pull. Thanks.

**Comment by taylortrimble** [comment]
After #158 and #163 I consider this ready to go. Good, good work! Thanks for getting it started. :beers:

**Comment by mplewis** [comment]
here goes nothing. :key:

**Pull request merged**